### PR TITLE
Fix (gptq): fix for act_order setup

### PR DESCRIPTION
**Pull request by Giuseppe5**

### CODE REVIEW

File: src/brevitas/graph/gptq.py [hunk 1]
```diff
@@ -264,10 +262,16 @@ def single_layer_update(self, percdamp=.01):
             i2 = min(i1 + blocksize, self.columns)
             count = i2 - i1
 
-            weight1 = weight[:, i1:i2]
+            if len(permutation_list) == 1:
+                perm = permutation_list[0]
+                weight1 = weight[:, perm[i1:i2]]
```
Comment (Giuseppe5) [inline src/brevitas/graph/gptq.py:267]: weight1 -> weight_block

File: src/brevitas/graph/gptq.py [hunk 2]
```diff
@@ -209,35 +217,46 @@ def update_batch(self, module, input, out, current_layer):
...
+            # For groupwise convolution, these operations are groupwise so we iterate
             for i in range(self.groups):
+                # Compute 'dead' activations
                 dead = torch.diag(self.H[i, :, :]) == 0
                 self.H[i, dead, dead] = 1
+                # If the diagonal of activations is zero, we set the weight to zero
                 weight[i, dead] = 0
                 if self.act_order:
+                    # Re-order Hessian so that weights associated to
+                    # higher magnitude activations are first
```
Comment (Giuseppe5) [inline src/brevitas/graph/gptq.py:240]: *are quantized first

File: src/brevitas/graph/gptq.py [hunk 3]
```diff
@@ -147,12 +147,20 @@ def __init__(self, layer, name, num_blocks, act_order) -> None:
...
+
+        # Number of rows is equal to the output channels (OC)
         self.rows = weight.shape[0]
         self.columns = weight.shape[1]
+
+        # Define how many columns to update in each mini-block
+        self.blocksize = math.ceil(self.columns / self.num_blocks)
+
+        # Initialize Hessian matrix and counter
         self.H = torch.zeros((self.groups, self.columns, self.columns), device=dev)
```
Comment (volcacius) [inline src/brevitas/graph/gptq.py:164]: Let's add dtype too

File: src/brevitas/graph/gptq.py [hunk 4]
```diff
@@ -209,35 +217,46 @@ def update_batch(self, module, input, out, current_layer):
...
+        # List with permutation tensors for the Hessian and Weight matrix.
+        # If act_order is False, the tensors will be ordered indexes.
+        # For groupwise convolution, we have one tensor per group,
+        # thus len(permutation_list) is always equal to self.groups.
+        # We do not explicity permute the weight matrix, only the Hessian.
         permutation_list = []
         if self.groups > 1:
+            # For groupwise convolution, these operations are groupwise so we iterate
             for i in range(self.groups):
+                # Compute 'dead' activations
```
Comment (volcacius) [inline src/brevitas/graph/gptq.py:233]: what is dead? Rephrase

File: src/brevitas/graph/gptq.py [hunk 5]
```diff
@@ -260,49 +279,67 @@ def single_layer_update(self, percdamp=.01):
...
+            if self.groups == 1:
+                perm = permutation_list[0]
+                weight_block = weight[:, perm[i1:i2]]  # This creates a copy
+            else:
+                # For groups > 1, we permute each row independently
+                weight_block = torch.empty(weight.shape[0], count, device=dev)  # [OC, i2-i1]
+                for ii, perm in enumerate(permutation_list):
+                    weight_block[ii, :] = weight[ii, perm[i1:i2]]  # This creates a copy
 
+            Err1 = torch.zeros_like(weight_block)  # [OC, i2-i1]
```
Comment (Giuseppe5) [inline src/brevitas/graph/gptq.py:296]: Err1 -> Err_block
Hinv1 -> Hinv_block

**Review by volcacius — APPROVED** [review]

**Pull request merged**